test/calendar.ts

```diff
@@ -0,0 +1,97 @@
+        await grist.setCustomWidgetMapping('startDate', /From/);
+        await grist.setCustomWidgetMapping('endDate', /To/);
+        await grist.setCustomWidgetMapping('title', /Label/);
+        await grist.setCustomWidgetMapping('isAllDay', /IsFullDay/);
+    });
+
+    it('should create new event when new row is added', async function () {
+        await grist.sendActions([['AddRecord', 'Table1', -1, {
+            From: new Date('2023-08-03 13:00'),
+            To: new Date('2023-08-03 14:00'),
+            Label: "New Event",
+            IsFullDay: false
+        }]]);
+        const mappedObject = await grist.getCustomWidgetObject(getAbstractFromCalendarObject(1));
+        assert.deepEqual(mappedObject, {
+            title: "New Event",
+            startDate: new Date('2023-08-03 13:00').toString(),
+            endDate: new Date('2023-08-03 14:00').toString(),
+            isAllDay: false
+        })
+    });
+
+    it('should create new all day event when new row is added', async function () {
+        await grist.sendActions([['AddRecord', 'Table1', -1, {
+            From: new Date('2023-08-04 13:00'),
+            To: new Date('2023-08-04 14:00'),
+            Label: "All Day Event",
+            IsFullDay: true
+        }]]);
+        const mappedObject = await grist.getCustomWidgetObject(getAbstractFromCalendarObject(2));
+        assert.equal(mappedObject.title, "All Day Event");
+        assert.equal(mappedObject.isAllDay, true);
+        // Ignoring time component, because it's not important in full day events
+        assert.equal(new Date(mappedObject.startDate).toDateString(),
+          new Date('2023-08-04 00:00:00').toDateString());
+        assert.equal(new Date(mappedObject.endDate).toDateString(),
+          new Date('2023-08-04 00:00:00').toDateString());
+    });
+
+    it('should update event when table data is changed', async function () {
+        await grist.sendActions([['UpdateRecord', 'Table1', 1, {
+            From: new Date('2023-08-03 13:00'),
+            To: new Date('2023-08-03 15:00'),
+            Label: "New Event",
+            IsFullDay: false
+        }]]);
+        const mappedObject = await grist.getCustomWidgetObject(getAbstractFromCalendarObject(1));
+        assert.deepEqual(mappedObject, {
+            title: "New Event",
+            startDate: new Date('2023-08-03 13:00').toString(),
+            endDate: new Date('2023-08-03 15:00').toString(),
+            isAllDay: false
+        })
+    });
+
+    it('should remove event when row is deleted', async function () {
+        await grist.sendActions([['RemoveRecord', 'Table1', 1]]);
+        const mappedObject = await grist.getCustomWidgetObject(getAbstractFromCalendarObject(1));
+        assert.notExists(mappedObject);
+    });
+
+
+    //TODO: test adding new events and moving existing one on the calendar. ToastUI is not best optimalized for drag and drop tests in mocha and i cannot yet make it working correctly. 
```

Comment:
For dragging and dropping, have you tried using `driver.withActions`?

Reply:
yes. problem is not in actually dragging element, but rather to calculate to where it need to be drag.